File: c++/299. Bulls and Cows.cpp
class Solution {
public:
    string getHint(string secret, string guess) {
        int n = secret.size();
        int visit[n] = {0};
        int bull = 0, cow = 0;
        vector<vector<int>> vv(10, vector<int>());
        for (int i = 0; i < n; i++) {
            if (secret[i] == guess[i]) {
                bull++;
                visit[i] = 1;
            } else 
                vv[secret[i] - '0'].push_back(i);
        }
        
        for (int i = 0; i < n; i++) {
            if (visit[i] == 0) {
                char ch = guess[i];
                if (!vv[ch - '0'].empty()) {
                    cow++;
                    vv[ch - '0'].erase(vv[ch - '0'].begin());
                }
            }
        }
        
        char buf[10];
        sprintf(buf, "%dA%dB", bull, cow);
        return string(buf);
    }
};

File: go/src/leetcode/299. Bulls and Cows/299. Bulls and Cows.go
package main

import "fmt"

func getHint(secret string, guess string) string {
	var x, y int
	n := len(secret)
	m1 := make(map[byte][]int)
	m2 := make(map[byte][]int)
	for i := 0; i < n; i++ {
		if secret[i] == guess[i] {
			x++
		} else {
			m1[secret[i]] = append(m1[secret[i]], i)
			m2[guess[i]] = append(m2[guess[i]], i)
		}
	}

	for d, p2 := range m2 {
		if p1, ok := m1[d]; ok {
			y += min(len(p1), len(p2))
		}
	}

	return fmt.Sprintf("%dA%dB", x, y)
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	fmt.Println(getHint("1807", "7810"))
	fmt.Println(getHint("1123", "0111"))
}